src/middlewares/rateLimit.ts

```diff
@@ -0,0 +1,58 @@
+import { Request, Response, NextFunction } from 'express'
+import { RequestersList } from '../utils'
+import { CONFIG as config } from '../config'
+import { sleep } from '../utils'
+import blackList from '../../blacklist.json'
+import spammerList from '../../spammerlist.json'
+
+// Export for testing
+export const requestersList = new RequestersList(blackList, spammerList)
+
+interface RpcRequest {
+  method: string
+  params: any[]
+}
+
+async function handleRejection(res: Response, softReject: boolean): Promise<void> {
+  if (softReject) {
+    const randomSleepTime = 10 + Math.floor(Math.random() * 10)
+    await sleep(randomSleepTime * 1000)
+    res.status(503).send('Network is currently busy. Please try again later.')
+  } else {
+    res.status(429).send('Rejected by rate-limiting')
+  }
+}
+
+async function checkRequest(ip: string, request: RpcRequest): Promise<boolean> {
+  return await requestersList.isRequestOkay(ip, request.method, request.params)
+}
+
+export async function rateLimitMiddleware(req: Request, res: Response, next: NextFunction) {
+  if (!config.rateLimit) {
+    next()
+    return
+  }
+  let ip = String(req.socket.remoteAddress)
```

Comment:
we should be using req.ip here @S0naliThakur !! 
By default, req.ip retrieves the IP address from the remoteAddress property of the underlying socket (TCP connection i.e. req.socket.remoteAddress
but it does more than that .. 
When an application is behind a reverse proxy (e.g., NGINX, HAProxy), the IP address in req.socket.remoteAddress is usually the proxy's IP.
To get the real client IP, the proxy includes an X-Forwarded-For header that contains the client's IP (and possibly other IPs in a chain).

If trust proxy is enabled in the Express app (app.set('trust proxy', true)), req.ip uses the X-Forwarded-For header to determine the client IP:
It extracts the first IP in the X-Forwarded-For header, as that represents the original client.
If X-Forwarded-For is absent or invalid, it falls back to req.socket.remoteAddress.

i've made a task to correct that.. https://linear.app/shm/issue/SHARD-1162/correction-of-rate-limiting-logic-in-json-rpc-server#comment-fead111a
but it also depends on this as well..